### PR TITLE
NCEI 2.0 Checker

### PR DESCRIPTION
**Pull request by Bobfrat**
@lukecampbell @mbiddle-nodc 
Adds NCEI 2.0 checks and tests. Still to do is the swath feature type detection and checker for 2.0.

### CODE REVIEW

File: cc_plugin_ncei/ncei_base.py [hunk 1]
```diff
@@ -557,50 +541,254 @@ def check_global_attributes(self, dataset):
...
+        sea_name = sea_name.replace(', ', ',')
+        sea_name = sea_name.split(',') if sea_name else []
+        for sea in sea_name:
+            recommended_ctx.assert_true(
+                sea.lower() in sea_names,
+                'sea_name attribute should exist and should be from the NODC sea names list: {} is not a valid sea name'.format(sea)
+            )
+
+        # Source: http://www.pelagodesign.com/blog/2009/05/20/iso-8601-date-validation-that-doesnt-suck/
+        iso8601_regex = r'^([\+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24\:?00)([\.,]\d+(?!:))?)?(\17[0-5]\d([\.,]\d+)?)?([zZ]|([\+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$'
```
Comment (lukecampbell) [inline cc_plugin_ncei/ncei_base.py]: could you replace this regex with something like https://github.com/ioos/compliance-checker/blob/master/compliance_checker/util.py#L18 ?

Comment (lukecampbell) [inline cc_plugin_ncei/ncei_base.py]: I think I'm the one that used the insane regex but using a third party module that the other checkers used would be better if only for the consistency.

File: cc_plugin_ncei/ncei_base.py [hunk 2]
```diff
@@ -557,50 +541,254 @@ def check_global_attributes(self, dataset):
...
+        :keywords = "" ; //............................................ HIGHLY RECOMMENDED - A comma separated list of keywords coming from the keywords_vocabulary. (ACDD)
+        :Conventions = "CF-1.6, ACDD-1.3" ; //......................... HIGHLY RECOMMENDED    - A comma separated list of the conventions being followed. Always try to use latest version. (CF/ACDD)
+        '''
+        test_ctx = TestCtx(BaseCheck.HIGH, 'Required global attributes')
+
+        conventions = getattr(dataset, 'Conventions', '')
+        feature_type = getattr(dataset, 'featureType', '')
+
+        # Define conventions
+        accepted_conventions = ['CF-1.6, ACDD-1.3', 'CF-1.6,ACDD-1.3']
```
Comment (lukecampbell) [inline cc_plugin_ncei/ncei_base.py]: I don't think there's any doctrine on the order in which the conventions are listed or how much white space is allowed between delimiters. Could we do something like this instead?

```python
accepted_conventions = ['CF-1.6', 'ACDD-1.3']
dataset_conventions = conventions.replace(' ', '').split(',')
for accepted_convention in accepted_conventions:
    if accepted_convention not in dataset_conventions:
      result_fail
else:
    result_pass
```

Or something of the like, that way the check becomes agnostic to order and if someone mistakenly puts in like 30 spaces, which I've seen due to copy paste errors.

Comment (Bobfrat) [inline cc_plugin_ncei/ncei_base.py]: 👍 good call

**Comment by Bobfrat** [comment]
@lukecampbell I just pushed updates for your suggestions.

**Comment by lukecampbell** [comment]
Ok, I need to test this using the CChecker Web tomorrow

**Comment by lukecampbell** [comment]
For the check dealing with "Recommended attributes for platform variable" it is still checking the legacy attribute name `nodc_code`, instead of allowing the new name `ncei_code`. 

Example dataset: https://data.nodc.noaa.gov/thredds/dodsC/testdata/mbiddle/GOLD_STANDARD_NETCDF/2.0/NCEI_point_template_v2.0_2016-06-15_133828.496967.nc.html

```
Name                             Priority:     Score:Reasoning
--------------------------------------------------------------------------------
Recommended attributes for platform var:2:     2/ 3 : At least one attribute
                                                      should be defined to
                                                      identify the platform:
                                                      nodc_code, wmo_code,
                                                      imo_code, call_sign.
```

**Comment by lukecampbell** [comment]
But the platform defines an `ncei_code` in accordance with NCEI 2.0 checks

```
	char platform1(maxStrlen64) ;
		platform1:comment = "Data is not actually collected from this platform, this is an example." ;
		platform1:long_name = "cordell bank monitoring station" ;
		platform1:ncei_code = "FIXED PLATFORM, MOORINGS" ;
		platform1:ioos_code = "urn:ioos:station:NCEI:Mooring1" ;
```

**Comment by lukecampbell** [comment]
@mbiddle-nodc I have a question about the `comment` attribute. According to the 2.0 templates, it's recommended, which is a 2-point value on the compliance checker scale. What if there's no useful information to include about the variable?

When I'm making datasets metadata compliant I often find myself just throwing in a comment attribute with complete fluff just to get a perfect score.

**Comment by lukecampbell** [comment]
Also, let's drop the allowed names for elevation/depth. People are always going to have names that aren't in our list but are still correct. 

Right now we restrict the allowed names for height to a (rather small) set. In practice people are going to have probably on the scale of hundreds of different and absolutely ok names for elevation/height/depth/z.

@mbiddle-nodc would you concur?

For example: http://oos.soest.hawaii.edu/thredds/dodsC/pibhmc_bathy_5m_marpi
is flagged for having a bad name for elevation but its `elev` which is perfectly ok and fairly intuitive as to what it's for. Even if the name's not the `standard_name` and `units` make it abundantly clear.

```
	float elev(lat, lon) ;
		elev:long_name = "elevation" ;
		elev:_FillValue = NaNf ;
		elev:standard_name = "height" ;
		elev:units = "meters" ;
		elev:positive = "up" ;
```

**Comment by mbiddle-nodc** [comment]
I'm wondering if the comment attribute should just be ignored. If they have it, great! If not, it's not a problem. Would it be possible to have a note, but not deduct points?

**Comment by lukecampbell** [comment]
@mbiddle-nodc with regards to the `comment` attribute, my idea was if it exists, we just check that it's a string and it's not empty, and award points for for those _if it exists_ otherwise we don't check anything so it's a `0/0` check.

**Comment by lukecampbell** [comment]
Ok, other than those three things, it's great and functions exactly how I imagined it would. I just tested it on a rather wide range of data using the IOOS catalog and I agree with all the corrective actions it recommends.

**Comment by mbiddle-nodc** [comment]
`comment` attribute check sounds good to me. 

As for the allowed names for the elevation/depth variable, I don't think it should be checking for the name of the variable in the first place, unless I'm misunderstanding.  My guidance has been, call the variable whatever your like (1, 2, 3, x, y, z, joe, smith, etc) just let the attributes describe the contents.

**Comment by lukecampbell** [comment]
Ok we're all on the same page then.

**Comment by Bobfrat** [comment]
@lukecampbell I made the 3 updates: ncei_code check for v2.0 platform vars, only check comment if it exists, and remove the check on elevation variable name

**Pull request merged**